class Solution {
    public boolean checkValidString(String s) {
        int high = 0 , low = 0;

        for(char ch : s.toCharArray()){
            if(ch == '('){
                low++;
                high++;
            }

            else if(ch == ')'){
                low--;
                high--;
            }

            else{
                low--;

                high++;
            }

            low = Math.max(low , 0);

            if(high < 0)
                return false;
        }

        return low == 0;
    }
}